### ISSUE TITLE
NoMethodError if select_append-method is sent before i.e. left_join-method

### ISSUE THREAD

**Issue by wuarmin** [hanami/model#475]
Hey,
the method "all_addresses" of following Repository
```ruby
class AddressesRepository < Hanami::Repository
  associations do
    belongs_to :address_types, foreign_key: :address_types_id
  end

  def all_addresses
    addresses.
      select_append { `'just_a_test'`.as(:test) }.
      left_join(address_types).
      map_to(Addresses).to_a
  end

end
```
fails with
```txt
NoMethodError: undefined method `dataset' for nil:NilClass
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/attribute.rb:322:in `to_sql_name'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/attribute.rb:113:in `qualified'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-3.3.3/lib/rom/schema.rb:126:in `each'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-3.3.3/lib/rom/schema.rb:126:in `each'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/schema.rb:46:in `map'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/schema.rb:46:in `qualified'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/relation/reading.rb:109:in `qualified'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/association.rb:63:in `join'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/relation/reading.rb:915:in `__join__'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-sql-1.3.5/lib/rom/sql/relation/reading.rb:565:in `join'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-3.3.3/lib/rom/pipeline.rb:80:in `method_missing'
    /home/armin/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rom-repository-1.4.0/lib/rom/repository/relation_proxy.rb:324:in `method_missing'
    /home/armin/.devel/essence/essence_backend/lib/essence_backend/repositories/addresses_repository.rb:9:in `all_addresses'
    /home/armin/.devel/essence/essence_backend/spec/essence_backend/repositories/addresses_repository_spec.rb:33:in `block (2 levels) in <top (required)>'
```
If I refactor "all_addresses" to
```ruby
  def all_addresses
    addresses.
      left_join(address_types).
      select_append { `'just_a_test'`.as(:test) }.
      map_to(Addresses).to_a
  end
```
it works. This is a very simple-example, but the issue is, I have some very complex repos with i.e. a core_statement-method, which serves as a base-brick for further query-methods.

For now I fixed it with explicit joins:
```ruby
  def all_addresses
    addresses.
      select_append { `'just_a_test'`.as(:test) }.
      join(:address_types, id: root[:address_types_id].qualified).
      map_to(Addresses).to_a
  end
```

Best regards.

**Comment by mereghost** [comment]
I might be wrong but this seems to be related with `rom`'s auto schema inference. As soon as you add a new column that isn't on the table anymore and it tries to check it for association information, it blows up.

**Comment by mereghost** [comment]
OTOH probably @solnic or @flash-gordon can provide a more in-depth explanation.

**Comment by flash-gordon** [comment]
see https://github.com/rom-rb/rom-sql/issues/288

**Comment by jodosha** [comment]
Fixed by https://github.com/rom-rb/rom-sql/pull/292

This will be released with Hanami 2.0 alpha in more than a month.
